libs/ngxtension/linked-query-param/src/linked-query-param.ts

```diff
@@ -0,0 +1,414 @@
+import {
+	effect,
+	inject,
+	Injectable,
+	Injector,
+	runInInjectionContext,
+	signal,
+	untracked,
+	WritableSignal,
+} from '@angular/core';
+import { toSignal } from '@angular/core/rxjs-interop';
+import {
+	ActivatedRoute,
+	NavigationExtras,
+	Params,
+	Router,
+} from '@angular/router';
+import { assertInjector } from 'ngxtension/assert-injector';
+import { createNotifier } from 'ngxtension/create-notifier';
+import { distinctUntilKeyChanged, map } from 'rxjs';
+
+/**
+ * The type of the serialized value.
+ * After transforming the value before it is passed to the query param, this type will be used.
+ */
+type SerializeReturnType = string | number | boolean | null | undefined;
+
+/**
+ * These are the options that can be passed to the `linkedQueryParam` function.
+ * They are taken from the `NavigationExtras` type in the `@angular/router` package.
+ */
+type NavigateMethodFields = Pick<
+	NavigationExtras,
+	| 'queryParamsHandling'
+	| 'onSameUrlNavigation'
+	| 'replaceUrl'
+	| 'skipLocationChange'
+>;
+
+/**
+ * Service to coalesce multiple navigation calls into a single navigation event.
+ */
+@Injectable({ providedIn: 'root' })
+export class LinkedQueryParamGlobalHandler {
+	private router = inject(Router);
+	/**
+	 * @internal
+	 * The current query params that will be set on the next navigation event.
+	 */
+	private _currentKeys: Record<string, SerializeReturnType> = {};
+	/**
+	 * @internal
+	 * The navigation extras that will be used on the next navigation event.
+	 */
+	private _navigationExtras: NavigationExtras = {};
+	/**
+	 * @internal
+	 * The notifier that will be used to schedule the navigation event.
+	 */
+	_schedulerNotifier = createNotifier();
+
+	constructor() {
+		effect(() => {
+			// listen to the scheduler notifier to schedule the navigation event
+			this._schedulerNotifier.listen();
+
+			// we need to untrack the navigation call in order to not register any other signal as a dependency
+			untracked(() => this.navigate().then());
+		});
+	}
+
+	/**
+	 * Sets the value of a query param.
+	 * This will be used on the next navigation event.
+	 */
+	setParamKeyValue(key: string, value: SerializeReturnType) {
+		this._currentKeys[key] = value;
+	}
+
+	/**
+	 * Sets the navigation extras that will be used on the next navigation event.
+	 */
+	setCurrentNavigationExtras(config: Partial<NavigateMethodFields>) {
```

Comment:
nit: if `config` isn't required as inferred by `config ?? {}` below, provide the default value to the parameter

```suggestion
	setCurrentNavigationExtras(config: Partial<NavigateMethodFields> = {}) {
```